File: AndroidArchitecture/mvvm-live/src/main/java/com/codearms/maoqiqi/app/Event.java
package com.codearms.maoqiqi.app;

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/21 17:18
 */
public class Event<T> {

    private T content;

    private boolean hasBeenHandled = false;

    public Event(T content) {
        if (content == null) {
            throw new IllegalArgumentException("null values in Event are not allowed.");
        }
        this.content = content;
    }

    public T getContentIfNotHandled() {
        if (hasBeenHandled) {
            return null;
        } else {
            hasBeenHandled = true;
            return content;
        }
    }

    public T getContent() {
        return content;
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/main/java/com/codearms/maoqiqi/app/Event.kt
package com.codearms.maoqiqi.app

/**
 * Used as a wrapper for data that is exposed via a LiveData that represents an event.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/21 17:18
 */
class Event<T>(private val content: T) {

    private var hasBeenHandled = false

    /**
     * Returns the content and prevents its use again.
     */
    fun getContentIfNotHandled(): T? {
        return if (hasBeenHandled) {
            null
        } else {
            hasBeenHandled = true
            content
        }
    }

    fun getContent(): T = content
}